keps/sig-cli/kubectl-apply-prune-enhancements.md

```diff
@@ -0,0 +1,273 @@
+
+#### How it works
+Since the apply-config object is not generated by `kubectl`, other tools or users need to generate an apply-config object configuration before piping them into `kubectl apply`. For example, in following workflows, `kustomize` or `helm` need to generate the apply-config object. Moreover, every time those tools run, it should generate the apply-config object with the same name.
+```
+kustomize build . | kubectl apply -f -
+# or
+helm template mychart | kubectl apply -f -
+```
+
+When `kubectl apply` is run, it looks for an apply-config object by label `kubectl.kubernetes.io/apply-cfg: "true"`. If no apply-config object is found, `kubectl` will give a warning that no apply-config object is found. When an apply-config object is found, the pruning action is triggered. If the label `kubectl.kubernetes.io/apply-hash` is present in the apply-config object, `kubectl` will validate this hash against the content stored in annotation. This validation is to prevent unintentional modification of the apply-config object annotations from users. Once the validation passes, the next step is to query the live apply-config object from the cluster. Assume the apply-config object has kind `K` and name `N`, `kubectl` looks for a live object from the cluster with the same kind and name.
+- If a live apply-config object is not found, Kubectl will create the apply-config object and the corresponding set of objects
+- If a live apply-config object is found, then the set of live objects will be found by reading the annotation field of the live apply-config object. Kubectl diffs the live set with new set of objects and takes following steps: 
+  - set the apply-config object annotation as the union of two sets and update the apply-config object
+  - create new objects
+  - prune old objects
+  - remove the pruned objects from apply-config object’s annotation field
+  - validate the `apply-hash` in apply-config object as a signal that pruning has successfully finished
+
+  
+pros:
+  - The apply-config object can be any type of objects
+  - No need to selecting matching objects by labels. More efficient
+  - Number of queries is linear to the number of resources being applied/pruned
+  
+cons:
+  - Users need be careful when modifying resource files so that the label `kubectl.kubernetes.io/apply-cfg: "true"` is not
+    removed unintentionally.
+  - Users/tools need to maintain the correct list of objects in annotations
+
+
+#### Risks and Mitigations
+
+Deciding the correct set of objects to change is hard. `--prune` Alpha tended to solve that problem in kubectl itself. In this new approach, kubectl is not going to decide such set. Instead it delegates the job to users or tools. Those who need to use `--prune` will be responsible to create an apply-config object as a record of the correct set of objects. Kubectl consumes this apply-config object and triggers the pruning action.
```

Comment:
In my view, it is actually kubectl's job to track the set, and it's not a good user experience for them to have to list their objects twice. The problem is maintaining identity between the prior and current invocations. I think the idea of storing a list of which objects were created is part of the right answer.

Perhaps instead of making the user write out a list, you can make the user include a placeholder object (and offer to create it)? Really the only thing you need the user to provide is a name which will stay stable between invocations. On every apply, you do something like:
1. get the current list of files ---> construct the new contents of the placeholder
2. look up the previous placeholder, which was stored as an API object.
3. diff, apply, prune
4. write the placeholder as an API object with the new contents

Reply:
So we agree on the idea of storing a list somewhere.

> Perhaps instead of making the user write out a list, you can make the user include a placeholder object (and offer to create it)? Really the only thing you need the user to provide is a name which will stay stable between invocations. 

Yes, we need a name staying stable. In addition to the name, we also need to know which types users don't want to prune, such as configmap/secret. Those types have to be listed out in the object storing the apply list. So the minimal information provided from users is an object name and a list of types. If users or tools can provide a list types that should be excluded from pruning, providing a full list of objects(excluding the types such as configmap/secret) doesn't cause much more work.